src/detect-engine-alert.c

```diff
@@ -337,72 +336,141 @@ static inline void FlowApplySignatureActions(
 }
 
 /**
- * \brief Check the threshold of the sigs that match, set actions, break on pass action
- *        This function iterate the packet alerts array, removing those that didn't match
- *        the threshold, and those that match after a signature with the action "pass".
- *        The array is sorted by action priority/order
+ * \brief Check the threshold of the sigs that match, check if tags should be applied,
+ * call functions to set actions to packet and flow
+ *
  * \param de_ctx detection engine context
  * \param det_ctx detection engine thread context
  * \param p pointer to the packet
+ * \param pa pointer to packet alert being processed
+ * \param s pointer to signature that generated alert
+ *
+ * \retval THRESHOLD_NOT_SUPPRESSED alert should be queued
+ * \retval THRESHOLD_SUPPRESSED alert is discarded
  */
-void PacketAlertFinalize(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx, Packet *p)
+static SigThresholdResults PacketAlertFinalize(DetectEngineCtx *de_ctx,
+        DetectEngineThreadCtx *det_ctx, Packet *p, PacketAlert *pa, const Signature *s)
 {
-    SCEnter();
+    SigThresholdResults res = PacketAlertHandle(de_ctx, det_ctx, s, p, pa);
+
+    if (res > THRESHOLD_DONT_ALERT) {
+        /* Now, if we have an alert, we have to check if we want
+         * to tag this session or src/dst host */
+        if (s->sm_arrays[DETECT_SM_LIST_TMATCH] != NULL) {
+            KEYWORD_PROFILING_SET_LIST(det_ctx, DETECT_SM_LIST_TMATCH);
+            SigMatchData *smd = s->sm_arrays[DETECT_SM_LIST_TMATCH];
+            while (1) {
+                /* tags are set only for alerts */
+                KEYWORD_PROFILING_START;
+                sigmatch_table[smd->type].Match(det_ctx, p, (Signature *)s, smd->ctx);
+                KEYWORD_PROFILING_END(det_ctx, smd->type, 1);
+                if (smd->is_last)
+                    break;
+                smd++;
+            }
+        }
 
-    /* sort the alert queue before thresholding and appending to Packet */
-    qsort(det_ctx->alert_queue, det_ctx->alert_queue_size, sizeof(PacketAlert),
-            AlertQueueSortHelper);
```

Comment:
It was suggested that we removed the qsort from here and called it after the alerts had been processed, but this leads to the current issue we're having with the sv test that is failing (exception-policy-default-01), that happens because the pass action is the last to be processed with the new approach, while in master it is processed first (so the flow and packet don't get the flag to be dropped set, and then things work fine).

Reply:
I think we need to do things in the correct order. If that requires sorting, so be it.